crates/kitsune-type/src/jsonld/serde/id.rs

```diff
@@ -0,0 +1,205 @@
+use core::{
+    fmt::{self, Formatter},
+    marker::PhantomData,
+};
+use serde::{
+    de::{
+        self, value::SeqAccessDeserializer, DeserializeSeed, Deserializer, IgnoredAny, MapAccess,
+        SeqAccess,
+    },
+    Deserialize,
+};
+
+/// Deserialises a single node identifier string or a set of node identifier strings.
+pub struct Id<T> {
+    seed: T,
+}
+
+struct Visitor<T>(T);
+
+#[cfg_attr(not(test), allow(dead_code))]
+impl<'de, T> Id<PhantomData<T>>
+where
+    T: Deserialize<'de>,
+{
+    pub fn new() -> Self {
+        Self::with_seed(PhantomData)
+    }
+
+    pub fn deserialize<D>(deserializer: D) -> Result<T, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Self::new().deserialize(deserializer)
+    }
+}
+
+impl<'de, T> Id<T>
+where
+    T: DeserializeSeed<'de>,
+{
+    pub fn with_seed(seed: T) -> Self {
+        Self { seed }
+    }
+}
+
+impl<'de, T> Default for Id<PhantomData<T>>
+where
+    T: Deserialize<'de>,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<'de, T> DeserializeSeed<'de> for Id<T>
+where
+    T: DeserializeSeed<'de>,
+{
+    type Value = T::Value;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_any(Visitor(self.seed))
+    }
+}
+
+impl<'de, T> de::Visitor<'de> for Visitor<T>
+where
+    T: DeserializeSeed<'de>,
+{
+    type Value = T::Value;
+
+    fn expecting(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str("a JSON-LD node")
+    }
+
+    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+    where
+        A: MapAccess<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(field_identifier, rename_all = "camelCase")]
```

Comment:
Not a review, just noting: Wow, this is really just an undocumented functionality, huh?